backend/internal/service/version_service.go

```diff
@@ -0,0 +1,80 @@
+package service
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync/atomic"
+	"time"
+)
+
+const versionCacheTTL = 15 * time.Minute
+
+type cacheEntry struct {
+	version     string
+	lastFetched time.Time
+}
+
+type VersionService struct {
+	httpClient *http.Client
+	cache      atomic.Value
+}
+
+func NewVersionService(httpClient *http.Client) *VersionService {
+	s := &VersionService{httpClient: httpClient}
+	s.cache.Store((*cacheEntry)(nil))
+	return s
+}
+
+// GetLatestVersion returns the latest available version from GitHub.
+// It caches the result for a short duration to avoid excessive API calls.
+func (s *VersionService) GetLatestVersion(ctx context.Context) (string, error) {
+	// Serve from cache if fresh
+	if entry, ok := s.cache.Load().(*cacheEntry); ok && entry != nil {
```

Comment:
While this will work fine, there's a situation where if multiple calls to GetLatestVersion start at the same time, the request is made multiple (unbounded) times in parallel.

An option could be to keep using an atomic for cache, but then adding a simple semaphore before starting the request. For example (with `s.requesting = make(chan struct{}, 1)`):

```go
// Serve from cache if fresh
// ... same as before

// Try to acquire the lock
// This blocks until another goroutine releases it
s.requesting <- struct{}{}
defer func () {
    // Release the lock
    <-s.requesting
} ()

// Now that we acquired the lock, check the cache again before making the request, as another goroutine may have populated it
// Same logic as above
```

Reply:
(I'm not going to say how "highly unlikely" generally happens with a large enough number of users... :) )

The request can take up to 5s, so there's a relatively large window of time where multiple requests can come at once. The biggest problem would be that with multiple requests in parallel, you may hit rate limiting in GitHub, causing responses to fail.